### ISSUE TITLE
AWS Signature V4 issue with space in object key

### ISSUE THREAD

**Issue by raelg** [http4k/http4k#656]
Hi, 

There is an issue with AwsAuth when objects keys have a space in them, which causes the CanonicalURI encoded with a `+` instead of `%20`, leading to an invalid signature:

https://github.com/http4k/http4k/blob/4c83d2e6c415fdbae466270861b6d5c377d313e8/http4k-aws/src/main/kotlin/org/http4k/aws/AwsCanonicalRequest.kt#L42

The [documentation](https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html) states the URL should be encoded according to [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986)

This is how the AWS Java SDK implements it:

https://github.com/aws/aws-sdk-java/blob/c639a88cc22b3282da450afec180be6ed08880e6/aws-java-sdk-core/src/main/java/com/amazonaws/util/SdkHttpUtils.java#L66

**Comment by raelg** [comment]
To further complicate this, I'm using a S3 clone (Dell ECS), which requires the path be URI-encoded twice, including the request URL  for GET and HEAD (in contrast to the [docs](https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html) which state _except for Amazon S3 which only gets URI-encoded once_). I'm not sure if this is true for AWS S3 too.

AWS provide an example implementation of UriEncode [here](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html), which converted to kotlin becomes:

```
fun uriEncode(input: CharSequence, encodeSlash: Boolean): String {
    val result = StringBuilder()
    for (element in input) {
        when (element) {
            in 'A'..'Z', in 'a'..'z', in '0'..'9', '_', '-', '~', '.' -> result.append(element)
            '/' -> result.append(if (encodeSlash) "%2F" else element)
            else -> result.append("%").append(Integer.toHexString(element.code))
        }
    }
    return result.toString()
}
```

**Comment by krissrex** [comment]
I wonder if I ran into the same bug, but with headers. 

I tried applying the OpenTelemetry tracing filter before the AWS filter, and my requests were rejected with 403 by AWS. Removing the OpenTelemetry filter solved it. This filter adds a header with the trace baggage. 

It used the `tracecontext,baggage` format.
I sadly don't have any examples of the values in the 3 headers (`traceparent`, `tracestate`, `baggage`).

**Comment by krissrex** [comment]
I made a test to check, and it does indeed fail:

```kotlin
    @Test
    fun `converts headers to canonical form`() {
        val request = Request(GET, "http://www.google.com/")
            .header("duplicate", "a")
            .header("duplicate", "b")
            .header("untrimmed", " surrounded with spaces ")
            .header("multi-spaces", "got  multiple   spaces      inside")
            .header("UPPERCASE", "x")
        val canonical = AwsCanonicalRequest.of(
            request, canonicalPayload
        )

    // assertThat...
```

Output:
```
GET
/

duplicate:a
duplicate:b
multi-spaces:got  multiple   spaces      inside
untrimmed:surrounded with spaces
uppercase:x

duplicate;duplicate;multi-spaces;untrimmed;uppercase
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```

The spec at https://docs.aws.amazon.com/IAM/latest/UserGuide/create-signed-request.html#create-canonical-request
states the following for canonical headers:
- > convert sequential spaces to a single space
- > separate the values for a multi-value header using commas

The headers `duplicate` and `multi-spaces` do not have correct values.

**Comment by krissrex** [comment]
That's not a regex parameter: 
![image](https://github.com/http4k/http4k/assets/7364831/05835187-5aaf-4973-b18b-682cd409326d)

:P

**Comment by s4nchez** [comment]
That's been fixed in 5.4.0.0. Thank you @krissrex for the PR!
